mpc_shifting_knapsack/shifting_knapsack.mpc

```diff
@@ -0,0 +1,150 @@
+import math
+
+
+program.use_edabit(True)
+
+
+def show_matrix(m):
+    for i in range(len(m)):
+        print_ln("%s ", m[i].reveal())
+    print_ln("")
+
+
+def knapsack(weights, values, C):
+    '''
+    The recursive param is to choose if solving the problem using the iterative
+    version or the recursive one.
+    '''
+    N = len(weights)
+    lwC = math.ceil(math.log2(C + 1)) + 1
+    ret = Array(1, sint)
+    ret[0] = 0
+
+    dp = Matrix(2, C + 1, sint)
+    dp.assign_all(0)
+
+    @for_range(N)
+    def _(i):
+        @for_range(C + 1)
+        def _(j):
+            dp[i % 2][j] = dp[(i + 1) % 2][j]
+
+        cyclic_shift(dp[(i + 1) % 2], C + 1 - weights[i], lwC)
+
+        @for_range(C + 1)
+        def _(j):
+            opt2 = dp[(i + 1) % 2][j] + values[i]
+            should_mov = (j >= weights[i]) * (opt2 > dp[i % 2][j])
+
+            dp[i % 2][j] = should_mov.if_else(opt2, dp[i % 2][j])
+
+        increased = dp[i % 2][C] > ret[0]
+
+        ret[0] = increased.if_else(dp[i % 2][C], ret[0])
+
+    return ret[0]
+
+
+def bitwise_xor(a, b):
+    bits_a = a.bit_decompose()
+    bits_b = b.bit_decompose()
+    bits_xor = Array(len(bits_a), sint)
+
+    for i in range(len(bits_a)):
```

Comment:
What if `len(bits_a) < len(bits_b)`? You will need to either note that the function assumes that `len(bits_a) >= len(bits_b)` or manually handle which of the bit arrays is longer.

Reply:
I think there is no need to check this because `a` and `b` are `sint` and `cint`. Hence, they will both have the same number of bits when you set the bit-length in the `compile.py` command. Both of them have the same representation. Does adding a comment about would help?

Reply:
Ahh. You are compiling this using `compile.py`. Then that means you can even set the bit length at the beginning of the current file as well, right?